sorts/binary_insertion_sort.py

```diff
@@ -0,0 +1,93 @@
+"""
+A pure Python implementation of the binary insertion sort algorithm
+
+Binary insertion sort is a sorting algorithm which is similar to the insertion sort,
+but instead of using linear search to find the location where an element should be
+inserted, we use binary search. Thus, we reduce the comparative value of inserting
+a single element from O (N) to O (log N).
+
+For doctests run following command:
+python3 -m doctest -v binary_insertion_sort.py
+
+For manual testing run:
+python3 binary_insertion_sort.py
+"""
+
+
+def binary_search(collection: list, val: int, start: int, end: int) -> int:
+    """
+    A pure Python implementation of the binary search  algorithm.
+
+    :param collection: some ascending sorted collection with comparable items
+    :param item: item value to search
+    :return: index of found item
+
+    Examples:
+    >>> binary_search([0, 5, 7, 10, 15], 0, 0, 4)
+    0
+
+    >>> binary_search([0, 5, 7, 10, 15], 15, 0, 4)
+    5
+
+    >>> binary_search([0, 5, 7, 10, 15], 5, 0, 4)
+    2
+
+    """
+    if start == end:
+        if collection[start] > val:
+            return start
+        else:
+            return start + 1
+    if start > end:
+        return start
+    mid = (start + end) // 2
+    if collection[mid] < val:
+        return binary_search(collection, val, mid + 1, end)
+    elif collection[mid] > val:
+        return binary_search(collection, val, start, mid - 1)
+    else:
+        return mid
+
+
+def binary_insertion_sort(collection: list) -> list:
+    """A pure Python implementation of the binary insertion sort algorithm
+
+    :param collection: some mutable ordered collection with heterogeneous
```

Comment:
The docstring is a direct copy paste from `sorts/insertion_sort.py`. Please update it as per the function.